### PR TITLE
[hot_reload] Add instance fields

### LINKED ISSUE #63643: [mono] Hot Reload: support adding instance fields
There are two separate ideas here:

1. To implement the `AddInstanceFieldToExistingType` EnC capability, we need to be able to hang new instance variables on existing classes (a prototype of the managed part is in Mono.HotReload.InstanceFieldTable in CoreLib).
2. Implement adding instance fields to newly-added classes.  This is part of the `NewTypeDefinition` capability.  Here we would like to allocate the storage for the new instance fields in the newly laid out class.  Otherwise even though we say we support `NewTypeDefinition`, we don't really.

Contributes to https://github.com/dotnet/runtime/issues/57365

### PR DESCRIPTION
**Pull request by lambdageek**
Fixes https://github.com/dotnet/runtime/issues/63643

Remaining work:
- [x] get/set fields in the debugger
- [x] Add instance properties
- [x] verify `ves_icall_property_info_get_default_value`works with added props
- [x] Add instance events (need added skeletons)
- [x] `mono_class_get_field_token`
- [x] `mono_class_get_event_token`
- [x] `mono_class_get_property_token`
- [x] `mono_field_get_rva`
- [x] `init_weak_fields_inner`
- [x] `ves_icall_RuntimeFieldInfo_GetFieldOffset`
- [x] `ves_icall_System_RuntimeFieldHandle_GetValueDirect`/`SetValueDirect`
- [x] audit callers of `mono_handle_unsafe_field_addr`
- [x] `mono_metadata_field_info_full`
- [x] `mono_metadata_get_marshal_info`, possibly

### CODE REVIEW

**Comment by lambdageek** [comment]
@thaystg @BrzVlad could you take a look.  Vlad: I'm adding an interp opcode and some transform changes to support it

**Comment by lambdageek** [comment]
(Some of the code is part of #76618 )

File: src/mono/mono/mini/interp/interp.c [hunk 1]
```diff
@@ -7322,7 +7322,20 @@ MINT_IN_CASE(MINT_BRTRUE_I8_SP) ZEROP_SP(gint64, !=); MINT_IN_BREAK;
...
+		MINT_IN_CASE(MINT_METADATA_UPDATE_LDFLDA) {
+			gpointer *dest = (gpointer*)(locals + ip [1]);
+			MonoObject *inst = LOCAL_VAR (ip [2], MonoObject*);
+			MonoType *field_type = frame->imethod->data_items [ip [3]];
+			uint32_t fielddef_token = GPOINTER_TO_UINT32 (frame->imethod->data_items [ip [4]]);
+			// FIXME: can we emit a call directly instead of a runtime-invoke?
+			gpointer field_addr = mono_metadata_update_added_field_ldflda (inst, field_type, fielddef_token, error);
+			/* FIXME: think about pinning the FieldStore and adding a second opcode to
+			 * unpin it */
+			mono_gc_wbarrier_generic_store_internal (dest, field_addr);
```
Comment (BrzVlad) [inline src/mono/mono/mini/interp/interp.c:7334]: This wbarrier is unnecessary. You can write the result like this: `LOCAL_VAR (ip [1], gpointer) = field_addr;`

**Comment by lambdageek** [comment]
Found a few more items that need work before this is ready; updated the PR description

**Comment by lambdageek** [comment]
Interesting. CoreCLR is crashing on something that I'm doing in the `AddInstanceField` test that I added in this PR :grin:

**Comment by lambdageek** [comment]
CoreCLR failure is https://github.com/dotnet/runtime/issues/76702

**Comment by lambdageek** [comment]
@thaystg @fanyang-mono This PR is essentially ready (I want to add a few more test cases for properties and events, but it's working locally, so I don't expect any other changes).  Could you please take another look.

File: src/mono/wasm/debugger/DebuggerTestSuite/HotReloadTests.cs [hunk 2]
```diff
@@ -561,5 +561,47 @@ public async Task DebugHotReloadMethod_AddingNewMethodWithoutAnyOtherChange_With
...
+            string asm_file_hot_reload = Path.Combine (DebuggerTestAppPath, "..", "wasm", asm_name+ ".dll");
+            var pause_location = await LoadAssemblyAndTestHotReload(asm_file, pdb_file, asm_file_hot_reload, "AddInstanceFields", "StaticMethod1",
+                                                                    expectBpResolvedEvent: false, sourcesToWait: new string [] { "MethodBody2.cs" });
+            var frame = pause_location["callFrames"][0];
+            var locals = await GetProperties(frame["callFrameId"].Value<string>());
+            await CheckObject(locals, "c", "ApplyUpdateReferencedAssembly.AddInstanceFields.C");
+            await SendCommandAndCheck (JObject.FromObject(new { }), "Debugger.resume", script_loc: null, line: -1, column: -1, function_name: null,
+                                       locals_fn: async (locals) => {
+                                           await CheckObject(locals, "c", "ApplyUpdateReferencedAssembly.AddInstanceFields.C");
+#if true
```
Comment (thaystg) [inline src/mono/wasm/debugger/DebuggerTestSuite/HotReloadTests.cs:580]: Why do we need this `#if true`?

Comment (lambdageek) [reply src/mono/wasm/debugger/DebuggerTestSuite/HotReloadTests.cs]: sorry. removed.

**Review by thaystg — APPROVED** [review]

**Pull request merged**